File: platform/app/api/agent_skills.py
"""Agent Skills 管理 API — 上传、列表、启停与文件预览。"""

from __future__ import annotations

import uuid
from typing import Annotated

from fastapi import APIRouter, Depends, File, Form, Query, UploadFile
from fastapi.responses import Response
from sqlalchemy.orm import Session

from app.api.deps import get_client_ip, require_permission
from app.database import get_db
from app.models.org import User
from app.schemas.agent_profile import AgentProfileDetailOut, AgentProfileOut, AgentProfilePatchIn
from app.schemas.agent_skill import AgentSkillFileContentOut, AgentSkillFileUpdateIn
from app.schemas.agent_skill import (
    AgentMemoryOut,
    AgentMemoryUpdateIn,
    AgentToolOut,
    AgentSkillCatalogItemOut,
    AgentSkillCreateIn,
    AgentSkillDetailOut,
    AgentSkillFileContentOut,
    AgentSkillFileUpdateIn,
    AgentSkillSummaryOut,
    AgentSkillUpdateIn,
    AgentSkillUploadOut,
    BuiltinSkillPatchIn,
    SkillInvokeIn,
    SkillInvokeOut,
    UnifiedSkillOut,
)
from app.schemas.common import ApiResponse, PageResult
from app.services import agent_skill_service as svc
from app.services import agent_profile_service as agent_profile_svc
from app.services import skill_registry_service as registry_svc
from app.services.audit_service import write_audit

router = APIRouter(
    prefix="/admin/agent-skills",
    tags=["admin", "agent-skills"],
    dependencies=[Depends(require_permission("feature.agent_skills"))],
)


@router.get("", response_model=ApiResponse[PageResult[AgentSkillSummaryOut]])
def list_agent_skills(
    db: Annotated[Session, Depends(get_db)],
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    q: str | None = Query(None, description="按 name / description 搜索"),
    enabled: bool | None = Query(None, description="按启用状态筛选"),
) -> ApiResponse[PageResult[AgentSkillSummaryOut]]:
    return ApiResponse(
        data=svc.list_skills(db, page=page, page_size=page_size, q=q, enabled_only=enabled)
    )


@router.get("/catalog", response_model=ApiResponse[list[AgentSkillCatalogItemOut]])
def read_skill_catalog(
    db: Annotated[Session, Depends(get_db)],
) -> ApiResponse[list[AgentSkillCatalogItemOut]]:
    """轻量目录（内置 + 上传），供智能体 Discovery 阶段使用。"""
    return ApiResponse(data=svc.get_skill_catalog(db, admin_view=True))


@router.get("/registry", response_model=ApiResponse[list[UnifiedSkillOut]])
def read_skill_registry(
    db: Annotated[Session, Depends(get_db)],
    include_disabled: bool = Query(True),
    catalog_only: bool = Query(True),
) -> ApiResponse[list[UnifiedSkillOut]]:
    """Skill 注册表：可对话调用的内置 Skill + 上传包（不含系统功能占位）。"""
    return ApiResponse(
        data=registry_svc.list_unified_skills(
            db, include_disabled=include_disabled, catalog_only=catalog_only
        )
    )


@router.get("/tools", response_model=ApiResponse[list[AgentToolOut]])
def read_agent_tools(
    db: Annotated[Session, Depends(get_db)],
    user: Annotated[User, Depends(require_permission("feature.agent_skills"))],
) -> ApiResponse[list[AgentToolOut]]:
    """平台 Agent 工具目录（只读，不可扩充）。"""
    from app.services.agent_tool_registry import list_agent_tools

    return ApiResponse(data=list_agent_tools(db, user=user))


@router.patch("/builtin/{skill_name}", response_model=ApiResponse[UnifiedSkillOut])
def patch_builtin_skill(
    skill_name: str,
    body: BuiltinSkillPatchIn,
    user: Annotated[User, Depends(require_permission("admin.user"))],
    db: Annotated[Session, Depends(get_db)],
    client_ip: Annotated[str | None, Depends(get_client_ip)] = None,
) -> ApiResponse[UnifiedSkillOut]:
    result = registry_svc.patch_builtin_skill(db, skill_name, enabled=body.enabled)
    write_audit(
        db,
        user_id=user.id,
        action="agent_skill.builtin_patch",
        resource_type="agent_skill",
        detail={"name": skill_name, "enabled": body.enabled},
        ip_address=client_ip,
    )
    return ApiResponse(data=result)


@router.post("/invoke", response_model=ApiResponse[SkillInvokeOut])
async def invoke_agent_skill(
    body: SkillInvokeIn,
    user: Annotated[User, Depends(require_permission("admin.user"))],
    db: Annotated[Session, Depends(get_db)],
    client_ip: Annotated[str | None, Depends(get_client_ip)] = None,
) -> ApiResponse[SkillInvokeOut]:
    """试调用 skill 工具（管理员调试，后续 agent 循环复用同一 executor）。"""
    result = await registry_svc.invoke_skill(
        db,
        user,
        skill_name=body.skill_name,
        tool_name=body.tool_name,
        params=body.params,
    )
    write_audit(
        db,
        user_id=user.id,
        action="agent_skill.invoke",
        resource_type="agent_skill",
        detail={
            "skill_name": body.skill_name,
            "tool_name": body.tool_name,
            "ok": result.ok,
        },
        ip_address=client_ip,
    )
    return ApiResponse(data=result)


@router.get("/agents", response_model=ApiResponse[list[AgentProfileOut]])
def list_system_agents(
    db: Annotated[Session, Depends(get_db)],
) -> ApiResponse[list[AgentProfileOut]]:
    """系统内置专精智能体列表（不可手动创建）。"""
    return ApiResponse(data=agent_profile_svc.list_agent_profiles(db))


@router.get("/agents/{agent_id}", response_model=ApiResponse[AgentProfileDetailOut])
def read_system_agent(
    agent_id: str,
    db: Annotated[Session, Depends(get_db)],
) -> ApiResponse[AgentProfileDetailOut]:
    return ApiResponse(data=agent_profile_svc.get_agent_profile_detail(db, agent_id))


@router.get(
    "/agents/{agent_id}/files/{file_path:path}",
    response_model=ApiResponse[AgentSkillFileContentOut],
)
def read_system_agent_file(
    agent_id: str,
    file_path: str,
    db: Annotated[Session, Depends(get_db)],
) -> ApiResponse[AgentSkillFileContentOut]:
    return ApiResponse(data=agent_profile_svc.get_agent_config_file(db, agent_id, file_path))


@router.put(
    "/agents/{agent_id}/files/{file_path:path}",
    response_model=ApiResponse[AgentSkillFileContentOut],
)
def update_system_agent_file(
    agent_id: str,
    file_path: str,
    body: AgentSkillFileUpdateIn,
    user: Annotated[User, Depends(require_permission("admin.user"))],
    db: Annotated[Session, Depends(get_db)],
    client_ip: Annotated[str | None, Depends(get_client_ip)] = None,
) -> ApiResponse[AgentSkillFileContentOut]:
    result = agent_profile_svc.update_agent_config_file(
        db, agent_id, file_path, body.content
    )
    write_audit(
        db,
        user_id=user.id,
        action="agent_profile.file_update",
        resource_type="agent_profile",
        detail={"agent_id": agent_id, "path": file_path},
        ip_address=client_ip,
    )
    return ApiResponse(data=result)


@router.patch("/agents/{agent_id}", response_model=ApiResponse[AgentProfileOut])
def patch_system_agent(
    agent_id: str,
    body: AgentProfilePatchIn,
    user: Annotated[User, Depends(require_permission("admin.user"))],
    db: Annotated[Session, Depends(get_db)],
    client_ip: Annotated[str | None, Depends(get_client_ip)] = None,
) -> ApiResponse[AgentProfileOut]:
    result = agent_profile_svc.patch_agent_profile(
        db,
        agent_id,
        enabled=body.enabled,
        skill_names=body.skill_names,
    )
    write_audit(
        db,
        user_id=user.id,
        action="agent_profile.patch",
        resource_type="agent_profile",
        detail={"agent_id": agent_id, **body.model_dump(exclude_unset=True)},
        ip_address=client_ip,
    )
    return ApiResponse(data=result)


@router.get("/{skill_id}", response_model=ApiResponse[AgentSkillDetailOut])
def read_agent_skill(
    skill_id: uuid.UUID,
    db: Annotated[Session, Depends(get_db)],
) -> ApiResponse[AgentSkillDetailOut]:
    return ApiResponse(data=svc.get_skill(db, skill_id))


@router.get(
    "/{skill_id}/files/{file_path:path}",
    response_model=ApiResponse[AgentSkillFileContentOut],
)
def read_agent_skill_file(
    skill_id: uuid.UUID,
    file_path: str,
    db: Annotated[Session, Depends(get_db)],
) -> ApiResponse[AgentSkillFileContentOut]:
    return ApiResponse(data=svc.get_skill_file_content(db, skill_id, file_path))


@router.get("/{skill_id}/download")
def download_agent_skill_zip(
    skill_id: uuid.UUID,
    db: Annotated[Session, Depends(get_db)],
) -> Response:
    """下载已安装 skill 的 ZIP 包（含目录结构，便于备份或迁移）。"""
    data, filename = svc.export_skill_zip(db, skill_id)
    return Response(
        content=data,
        media_type="application/zip",
        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
    )


@router.post("/generate", response_model=ApiResponse[AgentSkillSummaryOut])
def create_generated_agent_skill(
    body: AgentSkillCreateIn,
    user: Annotated[User, Depends(require_permission("admin.user"))],
    db: Annotated[Session, Depends(get_db)],
    client_ip: Annotated[str | None, Depends(get_client_ip)] = None,
) -> ApiResponse[AgentSkillSummaryOut]:
    result = svc.create_generated_skill(
        db,
        user,
        name=body.name,
        description=body.description,
        skill_md_body=body.skill_md_body,
        replace_existing=body.replace_existing,
    )
    write_audit(
        db,
        user_id=user.id,
        action="agent_skill.generate",
        resource_type="agent_skill",
        detail={"name": result.name},
        ip_address=client_ip,
    )
    return ApiResponse(data=result)


@router.put(
    "/{skill_id}/files/{file_path:path}",
    response_model=ApiResponse[AgentSkillFileContentOut],
)
def update_agent_skill_file(
    skill_id: uuid.UUID,
    file_path: str,
    body: AgentSkillFileUpdateIn,
    user: Annotated[User, Depends(require_permission("admin.user"))],
    db: Annotated[Session, Depends(get_db)],
    client_ip: Annotated[str | None, Depends(get_client_ip)] = None,
) -> ApiResponse[AgentSkillFileContentOut]:
    result = svc.update_skill_file(db, skill_id, file_path, body.content)
    write_audit(
        db,
        user_id=user.id,
        action="agent_skill.file_update",
        resource_type="agent_skill",
        detail={"skill_id": str(skill_id), "path": file_path},
        ip_address=client_ip,
    )
    return ApiResponse(data=result)


@router.post("/upload/zip", response_model=ApiResponse[AgentSkillUploadOut])
async def upload_agent_skill_zip(
    user: Annotated[User, Depends(require_permission("admin.user"))],
    db: Annotated[Session, Depends(get_db)],
    file: UploadFile = File(..., description="Skill ZIP（含 skill 目录）"),
    replace_existing: bool = Form(True),
    client_ip: Annotated[str | None, Depends(get_client_ip)] = None,
) -> ApiResponse[AgentSkillUploadOut]:
    content = await file.read()
    result = svc.upload_skill_zip(
        db, user, content, replace_existing=replace_existing
    )
    write_audit(
        db,
        user_id=user.id,
        action="agent_skill.upload_zip",
        resource_type="agent_skill",
        detail={
            "filename": file.filename,
            "bytes": len(content),
            "skills": [s.name for s in result.skills],
        },
        ip_address=client_ip,
    )
    return ApiResponse(data=result)


@router.post("/upload/folder", response_model=ApiResponse[AgentSkillUploadOut])
async def upload_agent_skill_folder(
    user: Annotated[User, Depends(require_permission("admin.user"))],
    db: Annotated[Session, Depends(get_db)],
    files: list[UploadFile] = File(..., description="文件夹内各文件的相对路径"),
    paths: Annotated[list[str] | str, Form(...)] = ...,
    replace_existing: bool = Form(True),
    client_ip: Annotated[str | None, Depends(get_client_ip)] = None,
) -> ApiResponse[AgentSkillUploadOut]:
    path_list = paths if isinstance(paths, list) else [paths]
    if len(files) != len(path_list):
        from app.core.exceptions import bad_request

        raise bad_request("files 与 paths 数量不一致")
    entries: list[tuple[str, bytes]] = []
    for upload, rel in zip(files, path_list, strict=True):
        entries.append(((rel or upload.filename or "").strip(), await upload.read()))
    result = svc.upload_skill_folder(
        db, user, entries, replace_existing=replace_existing
    )
    write_audit(
        db,
        user_id=user.id,
        action="agent_skill.upload_folder",
        resource_type="agent_skill",
        detail={
            "file_count": len(entries),
            "skills": [s.name for s in result.skills],
        },
        ip_address=client_ip,
    )
    return ApiResponse(data=result)


@router.patch("/{skill_id}", response_model=ApiResponse[AgentSkillSummaryOut])
def patch_agent_skill(
    skill_id: uuid.UUID,
    body: AgentSkillUpdateIn,
    user: Annotated[User, Depends(require_permission("admin.user"))],
    db: Annotated[Session, Depends(get_db)],
    client_ip: Annotated[str | None, Depends(get_client_ip)] = None,
) -> ApiResponse[AgentSkillSummaryOut]:
    result = svc.update_skill(
        db,
        skill_id,
        enabled=body.enabled,
        description=body.description,
    )
    write_audit(
        db,
        user_id=user.id,
        action="agent_skill.update",
        resource_type="agent_skill",
        detail={"skill_id": str(skill_id), **body.model_dump(exclude_unset=True)},
        ip_address=client_ip,
    )
    return ApiResponse(data=result)


@router.delete("/{skill_id}", response_model=ApiResponse[None])
def delete_agent_skill(
    skill_id: uuid.UUID,
    user: Annotated[User, Depends(require_permission("feature.agent_skills"))],
    db: Annotated[Session, Depends(get_db)],
    client_ip: Annotated[str | None, Depends(get_client_ip)] = None,
) -> ApiResponse[None]:
    skill = svc.get_skill(db, skill_id)
    svc.delete_skill(db, skill_id)
    write_audit(
        db,
        user_id=user.id,
        action="agent_skill.delete",
        resource_type="agent_skill",
        detail={"skill_id": str(skill_id), "name": skill.name},
        ip_address=client_ip,
    )
    return ApiResponse(data=None)

File: frontend/src/api/agentSkills.js
import { api, fetchWithTimeout, getApiBase, getToken } from "./http.js";
import { downloadBlob } from "../utils/downloadBlob.js";

export async function fetchAgentSkillRegistry({ includeDisabled = true, catalogOnly = true } = {}) {
  const params = new URLSearchParams();
  if (!includeDisabled) params.set("include_disabled", "false");
  if (!catalogOnly) params.set("catalog_only", "false");
  const qs = params.toString();
  return api(`/api/v1/admin/agent-skills/registry${qs ? `?${qs}` : ""}`);
}

export async function fetchAgentTools() {
  return api("/api/v1/admin/agent-skills/tools");
}

export async function patchBuiltinSkill(name, { enabled, title, description }) {
  const body = { enabled };
  if (title !== undefined) body.title = title;
  if (description !== undefined) body.description = description;
  return api(`/api/v1/admin/agent-skills/builtin/${encodeURIComponent(name)}`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
}

export async function invokeAgentSkill({ skillName, toolName = "search", params = {} }) {
  return api("/api/v1/admin/agent-skills/invoke", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      skill_name: skillName,
      tool_name: toolName,
      params,
    }),
  });
}

export async function fetchAgentSkills({ page = 1, pageSize = 15, q = "", enabled = null } = {}) {
  const params = new URLSearchParams({ page: String(page), page_size: String(pageSize) });
  if (q) params.set("q", q);
  if (enabled === true) params.set("enabled", "true");
  if (enabled === false) params.set("enabled", "false");
  return api(`/api/v1/admin/agent-skills?${params}`);
}

export async function fetchAgentSkill(skillId) {
  return api(`/api/v1/admin/agent-skills/${skillId}`);
}

export async function fetchAgentSkillFile(skillId, filePath) {
  const encoded = filePath.split("/").map(encodeURIComponent).join("/");
  return api(`/api/v1/admin/agent-skills/${skillId}/files/${encoded}`);
}

export async function downloadAgentSkillZip(skillId, filename = "skill.zip") {
  const headers = {};
  const token = getToken();
  if (token) headers.Authorization = `Bearer ${token}`;
  const res = await fetchWithTimeout(
    `${getApiBase()}/api/v1/admin/agent-skills/${skillId}/download`,
    { headers },
    60_000
  );
  if (!res.ok) {
    const json = await res.json().catch(() => ({}));
    throw new Error(json?.message || res.statusText || "下载失败");
  }
  const blob = await res.blob();
  downloadBlob(blob, filename);
}

export async function uploadAgentSkillZip(file, { replaceExisting = true } = {}) {
  const form = new FormData();
  form.append("file", file);
  form.append("replace_existing", replaceExisting ? "true" : "false");
  return api("/api/v1/admin/agent-skills/upload/zip", { method: "POST", body: form });
}

export async function uploadAgentSkillFolder(fileList, { replaceExisting = true } = {}) {
  const form = new FormData();
  for (const file of fileList) {
    form.append("files", file);
    const rel = file.webkitRelativePath || file.name;
    form.append("paths", rel);
  }
  form.append("replace_existing", replaceExisting ? "true" : "false");
  return api("/api/v1/admin/agent-skills/upload/folder", { method: "POST", body: form });
}

export async function updateAgentSkill(skillId, payload) {
  return api(`/api/v1/admin/agent-skills/${skillId}`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
}

export async function updateAgentSkillFile(skillId, filePath, content) {
  const encoded = filePath.split("/").map(encodeURIComponent).join("/");
  return api(`/api/v1/admin/agent-skills/${skillId}/files/${encoded}`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ content }),
  });
}

export async function createGeneratedAgentSkill({
  name,
  description,
  skillMdBody,
  replaceExisting = false,
}) {
  return api("/api/v1/admin/agent-skills/generate", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      name,
      description,
      skill_md_body: skillMdBody,
      replace_existing: replaceExisting,
    }),
  });
}

export async function fetchAgentMemory() {
  return api("/api/v1/ai-chat/agent-memory");
}

export async function updateAgentMemory(content) {
  return api("/api/v1/ai-chat/agent-memory", {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ content }),
  });
}

export async function clearAgentMemory() {
  return api("/api/v1/ai-chat/agent-memory", { method: "DELETE" });
}

export async function deleteAgentSkill(skillId) {
  return api(`/api/v1/admin/agent-skills/${skillId}`, { method: "DELETE" });
}

export async function fetchAgentProfiles() {
  return api("/api/v1/admin/agent-skills/agents");
}

export async function fetchRoutingSkillsMd() {
  return api("/api/v1/admin/agent-skills/routing/skills.md");
}

export async function fetchRoutingAgentsMd() {
  return api("/api/v1/admin/agent-skills/routing/agents.md");
}

export async function fetchAgentProfile(agentId) {
  return api(`/api/v1/admin/agent-skills/agents/${encodeURIComponent(agentId)}`);
}

export async function fetchAgentProfileFile(agentId, filePath) {
  const encoded = filePath.split("/").map(encodeURIComponent).join("/");
  return api(
    `/api/v1/admin/agent-skills/agents/${encodeURIComponent(agentId)}/files/${encoded}`
  );
}

export async function updateAgentProfileFile(agentId, filePath, content) {
  const encoded = filePath.split("/").map(encodeURIComponent).join("/");
  return api(
    `/api/v1/admin/agent-skills/agents/${encodeURIComponent(agentId)}/files/${encoded}`,
    {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ content }),
    }
  );
}

export async function patchAgentProfile(agentId, payload) {
  return api(`/api/v1/admin/agent-skills/agents/${encodeURIComponent(agentId)}`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
}

// ── 知识库文件夹挂载 ──────────────────────────────────

export async function fetchKnowledgeMounts(agentId) {
  return api(`/api/v1/admin/agent-skills/agents/${encodeURIComponent(agentId)}/knowledge-mounts`);
}

export async function addKnowledgeMount(agentId, { datasetId, folderId, scope, label }) {
  return api(`/api/v1/admin/agent-skills/agents/${encodeURIComponent(agentId)}/knowledge-mounts`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      dataset_id: datasetId,
      folder_id: folderId,
      scope,
      label: label || undefined,
    }),
  });
}

export async function removeKnowledgeMount(agentId, mountId) {
  return api(`/api/v1/admin/agent-skills/agents/${encodeURIComponent(agentId)}/knowledge-mounts/${mountId}`, {
    method: "DELETE",
  });
}
